### PR TITLE
P4 2533 improve import

### PR DESCRIPTION
**Pull request by jcavell**
Mega PR, sorry.

This improves the import in a number of ways:

**The import of person / profile is now separate to the import of moves, journeys, events.**
People and Profile reports are imported and persisted separately
This means that if an updated person is received it will result in the JPC data being updated.
PII data has been removed from Move. Move contains a nullable Person now.

**The separate move and report data structures have been consolidated**
There was unnecessary overlap, so a single entity is now used both for JPA and for Klaxon JSON parsing
Also, enums such as Supplier, MoveStatus and JourneyState are now referenced as such rather than as Strings

**The OneToMany relationships have been removed and are managed outside JPA**
JPA was being problematic and inefficient handling these, so it's now done manually

**Notes and comments are now truncated at 255 chars**
Some inserts were failing due to length of varchar being exceeded.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/Move.kt [hunk 1]
```diff
@@ -128,8 +134,31 @@ data class Move(
         fun toSiteName() = toSiteName ?: toNomisAgencyId
         fun toLocationType() = toLocationType?.name ?: "NOT MAPPED"
 
+        companion object{
+                private val dateFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy")
+                private val timeFormatter = DateTimeFormatter.ofPattern("HH:mm")
+
+                val CANCELLATION_REASON_CANCELLED_BY_PMU = "cancelled_by_pmu"
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/Move.kt:141]: `const val CANCELLATION_REASON_CANCELLED_BY_PMU = "cancelled_by_pmu"`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/JpcApplication.kt [hunk 2]
```diff
@@ -21,11 +20,11 @@ fun main(args: Array<String>) {
       (context.getBean(ManualLocationImporter::class) as ManualLocationImporter).let { SpringApplication.exit(context, it.import()) }
     }
 
-    context.environment.getProperty("import-serco-prices")?.let {
+    context.environment.getProperty("import-SERCO-prices")?.let {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/JpcApplication.kt]: We don't need any of this logic in this class now.  We can do what we need via Spring Shell.  Have I mentioned Spring Shell before? A rhetorical question.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/Event.kt [hunk 3]
```diff
@@ -53,22 +52,58 @@ data class Event constructor(
...
         fun fromJson(json: String): Event? {
             return Klaxon().
+            fieldConverter(SupplierParser::class, supplierConverter).
             fieldConverter(EventDateTime::class, dateTimeConverter).
             parse<Event>(json)
         }
     }
 
+    init {
+        notes = notes?.take(255)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/Event.kt:104]: Perhaps move this to the top of the class as in above the functions? (line 58)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/MoveFilterer.kt [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    /**
+     * All other completed moves not covered by standard, redirect, long haul or lockout moves
+     */
+    fun isMultiTypeMove(move: Move) = isCompleted(move) &&
+        !isStandardMove(move) &&
+        !isRedirectionMove(move) &&
+        !isLongHaulMove(move) &&
+        !isLockoutMove(move)
+
+/**Ω
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/MoveFilterer.kt:92]: Where did this character come from? :smile:

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/MoveFilterer.kt [hunk 5]
```diff
@@ -0,0 +1,114 @@
+package uk.gov.justice.digital.hmpps.pecs.jpc.importer.move
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/MoveFilterer.kt:1]: Package definition doesn't match that of the files actual location in the project. Also a few unused imports.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/JourneyStateParser.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+@Target(AnnotationTarget.FIELD)
+annotation class JourneyStateParser
+
+val journeyStateConverter = object: Converter {
+
+    override fun canConvert(cls: Class<*>) = cls == JourneyState::class.java
+
+    override fun fromJson(jv: JsonValue) = JourneyState.valueOfCaseInsensitive(jv.string)
+
+    override fun toJson(value: Any) = """"${value}""""
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/JourneyStateParser.kt:16]: Don't need curly braces.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/MoveStatusParser.kt [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+@Target(AnnotationTarget.FIELD)
+annotation class MoveStatusParser
+
+val moveStatusConverter = object: Converter {
+
+    override fun canConvert(cls: Class<*>) = cls == MoveStatus::class.java
+
+    override fun fromJson(jv: JsonValue) = MoveStatus.valueOfCaseInsensitive(jv.string)
+
+    override fun toJson(value: Any) = """"${value}""""
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/MoveStatusParser.kt:16]: Curly braces not needed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/Person.kt [hunk 8]
```diff
@@ -4,39 +4,59 @@ import com.beust.klaxon.Json
 import com.beust.klaxon.Klaxon
 import java.time.LocalDate
 import java.time.LocalDateTime
+import javax.persistence.Column
+import javax.persistence.Entity
+import javax.persistence.Id
+import javax.persistence.Table
 import javax.validation.constraints.NotBlank
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/Person.kt:11]: Redundant import.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/ReportParser.kt [hunk 9]
```diff
@@ -29,56 +31,58 @@ object ReportParser {
         }.filterNotNull()
     }
 
-    fun parseAsProfileIdToPersonId(profileFiles: List<String>): Map<String, String> {
+    fun parseAsPersonIdToProfileId(profileFiles: List<String>): Map<String, String> {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/ReportParser.kt]: Can be private.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/ReportParser.kt [hunk 10]
```diff
@@ -29,56 +31,58 @@ object ReportParser {
...
         logger.info("Parsing profiles")
-        return read(profileFiles) { Profile.fromJson(it) }.associateBy(keySelector = { it.id }, valueTransform = { it.personId })
+        return read(profileFiles) { Profile.fromJson(it) }.associateBy(keySelector = { it.personId }, valueTransform = { it.id })
     }
 
-    fun parseAsPersonIdToPerson(peopleFiles: List<String>): Map<String, Person> {
+    fun parseAsPerson(peopleFiles: List<String>): Sequence<Person> {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/ReportParser.kt:34]: Can be private.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/SupplierParser.kt [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+@Target(AnnotationTarget.FIELD)
+annotation class SupplierParser
+
+val supplierConverter = object: Converter {
+
+    override fun canConvert(cls: Class<*>) = cls == Supplier::class.java
+
+    override fun fromJson(jv: JsonValue) = Supplier.valueOfCaseInsensitive(jv.string)
+
+    override fun toJson(value: Any) = """"${value}""""
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/SupplierParser.kt:16]: Curly's not needed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/Move.kt [hunk 12]
```diff
@@ -139,53 +168,85 @@ data class Move(
...
+
+        fun hasAllOf(vararg ets: EventType) = getEvents(*ets).size == ets.size
+        fun hasAnyOf(vararg ets: EventType) = getEvents(*ets).isNotEmpty()
+        fun hasNoneOf(vararg ets: EventType) = !hasAnyOf(*ets)
+
+        fun getEvents(vararg ets: EventType) =
+                this.events.filter { ets.map{it.value}.contains(it.type) } +
+                        this.journeys.flatMap { it.events }.filter{ ets.map{it.value}.contains(it.type)  }
+
+        init {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/Move.kt:233]: Again perhaps move above functions to top of class.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/MovePersister.kt [hunk 13]
```diff
@@ -4,151 +4,114 @@ import org.slf4j.LoggerFactory
...
+                        },
+                        notes = moveEvents.notes(),
+                )
+
+                moveRepository.save(moveToSave)
+
+                val journeysToSave = processJourneys(moveToSave, journeys, journeyEvents)
+                journeyRepository.saveAll(journeysToSave)
+
+                eventRepository.saveAll(moveEvents + journeyEvents)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/MovePersister.kt:62]: Does this need to be included in the flush as well?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/MovePersister.kt [hunk 14]
```diff
@@ -4,151 +4,114 @@ import org.slf4j.LoggerFactory
...
+                journey.copy(
+                    events = listOf(),
+                    pickUpDateTime = pickUp,
+                    dropOffDateTime = dropOff,
+                    effectiveYear = pickUp?.year ?: effectiveYearForDate(move.moveDate ?: timeSource.date())
+                )
+            }
     }
 
+    fun fakeCancelledJourney(move: Move): Journey {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/MovePersister.kt:94]: Probably warrants some documentation as to why we need this.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/MoveRepository.kt [hunk 15]
```diff
@@ -1,5 +1,9 @@
 package uk.gov.justice.digital.hmpps.pecs.jpc.move
 
 import org.springframework.data.jpa.repository.JpaRepository
+import java.util.*
 
-interface MoveRepository : JpaRepository<Move, String>
\ No newline at end of file
+interface MoveRepository : JpaRepository<Move, String>{
+
+    fun findByProfileId(profileId : String): Optional<Move>
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/MoveRepository.kt:8]: Couldn't find anywhere this is used....

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/PersonPersister.kt [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+    fun persistPeople(people: List<Person>) {
+        logger.info("Persisting ${people.size} people")
+        people.forEach { person ->
+            Result.runCatching { personRepository.save(person) }
+        }
+    }
+
+    fun persistProfiles(profiles: List<Profile>) {
+        logger.info("Persisting ${profiles.size} people")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/PersonPersister.kt:24]: `logger.info("Persisting ${profiles.size} profiles")` (not people).

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/PersonPersister.kt [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+class PersonPersister(private val personRepository: PersonRepository,
+                      private val profileRepository: ProfileRepository,
+                      private val timeSource: TimeSource) {
+
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    fun persistPeople(people: List<Person>) {
+        logger.info("Persisting ${people.size} people")
+        people.forEach { person ->
+            Result.runCatching { personRepository.save(person) }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/move/PersonPersister.kt:19]: Perhaps log failures.   Same for profiles below.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/MoveService.kt [hunk 18]
```diff
@@ -9,19 +9,22 @@ import java.time.LocalDate
 
 
 @Service
-class MoveService(private val moveQueryRepository: MoveQueryRepository, private val eventRepository: EventRepository) {
+class MoveService(private val moveQueryRepository: MoveQueryRepository,
+                  private val journeyRepository: JourneyRepository,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/MoveService.kt:13]: JourneyRepository not used.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/MoveService.kt [hunk 19]
```diff
@@ -9,19 +9,22 @@ import java.time.LocalDate
...
+                  private val eventRepository: EventRepository) {
 
     private val logger = LoggerFactory.getLogger(javaClass)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/MoveService.kt:16]: Logger not used.

File: src/main/resources/application.yml [hunk 20]
```diff
@@ -5,7 +5,7 @@ spring:
     database: POSTGRESQL
     show-sql: false
     hibernate:
-      ddl-auto: update
+      ddl-auto: create
```
Comment (matt-crowson-moj) [inline src/main/resources/application.yml:8]: :fearful:

File: src/test/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/ReportImporterTest.kt [hunk 21]
```diff
@@ -44,46 +46,48 @@ internal class ReportImporterTest(@Autowired provider: ReportingProvider, @Autow
...
 
-    @Test
-    fun `Person populated correctly`(){
-        // This should pick up the latest person file
-        val person = moves.toList()[0].person
-        assertThat(person?.dateOfBirth).isEqualTo(LocalDate.of(1980, 12, 25))
-        assertThat(person?.prisonNumber).isEqualTo("PRISON1")
-
-    }
+//    @Test
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/ReportImporterTest.kt:49]: Can this be removed?

**Review by matt-crowson-moj — CHANGES_REQUESTED** [review]
I think there a few things that probably need addressing.  Mainly small stuff though.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/MoveFilterer.kt (hunk 5 above)
Comment (jcavell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/importer/report/MoveFilterer.kt:1]: Yes agree, I think I might address that in a separate PR

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**